### ISSUE TITLE
ENH: add method to rv_generic to return Fisher information matrix

### ISSUE THREAD

**Issue by fancidev** [scipy/scipy#19912]
### Is your feature request related to a problem? Please describe.

After fitting a distribution to data by MLE, I would like to know the (asymptotic) variance of the estimator to get a sense of its accuracy. This is given by the inverse of the Fisher (observed) information matrix. I don’t find how to do that in scipy.

### Describe the solution you'd like.

Add a `fisher` method to rv_generic to return the Fisher information or observed information matrix.

The signature could be the following:

`def fisher(self, theta, data=None)` returns the Fisher information matrix evaluated at `theta` if data is None, or the observed information matrix if data is not None.

The generic implementation could do a numerical approximation of the Hessian for observed information, and raise an exception if data is None.

### Describe alternatives you've considered.

For my particular task at hand, I approximated the Hessian by numerical difference. It’s also possible to code the analytical formula for the distribution of interest if an analytical formula exists.

**Comment by dschmitz89** [comment]
Adding Fisher information to the current distribution infrastructure would be hard to do. But I think it could be incorporated into the new one which is currently in the making (#15928)

**Comment by mdhaber** [comment]
This is part of why I wrote [`_differentiate`](https://github.com/scipy/scipy/pull/18811). It needs to be extended substantially (multiple dimensions and higher integrals), but the ball is rolling.

**Comment by fancidev** [comment]
> This is part of why I wrote [`_differentiate`](https://github.com/scipy/scipy/pull/18811). It needs to be extended substantially (multiple dimensions and higher integrals), but the ball is rolling.

Nice stuff. Seems non-trivial to extend `_differentiate` to support second (partial) derivatives?

**Comment by mdhaber** [comment]
Well, trivial to extend it by differentiating the derivative, but non-trivial to do the right thing. But I familiarized myself with the theory and how to make it adaptive. So at some point this will happen.

**Comment by mdhaber** [comment]
Now with `scipy.differentiate.hessian`, it's pretty simple to get reasonably accurate results in many cases. Here's a demonstration for several discrete and continuous distributions, comparing the results against those preduced by `jax.hessian`.

```python3
import numpy as np
import jax
from scipy import stats, differentiate 

rng = np.random.default_rng(2549824598234528)

dists = [('bernoulli', dict(p=0.3)),
         ('beta', dict(a=1.0, b=2.0, scale=3.0)),
         ('cauchy', dict(loc=1.0, scale=2.0)),
         ('chi2', dict(df=5., scale=2.0)),
         ('expon', dict(scale=3.)),
         ('gamma', dict(a=3., scale=2.)),
         ('gennorm', dict(beta=1.4)),
         ('geom', dict(p=0.7)),
         ('laplace', dict(loc=1., scale=2.)),
         ('logistic', dict(loc=1., scale=2.)),
         ('nbinom', dict(n=2., p=0.3)),
         ('norm', dict(loc=1., scale=2.)),
         ('poisson', dict(mu=1.5)),
         ('vonmises', dict(kappa=1.5))]

for distinfo in dists:
    distname, params = distinfo
    pnames = list(params.keys())
    p = list(params.values())
    dist = getattr(stats, distname)
    jaxdist = getattr(jax.scipy.stats, distname)

    n = 10
    x = dist.rvs(**params, size=n, random_state=rng)

    def llf(p):
        p = [pi[..., np.newaxis] for pi in p]
        params = dict(zip(pnames, p))
        try:
            return -np.sum(dist.logpdf(x, **params), axis=-1)
        except AttributeError:
            return -np.sum(dist.logpmf(x, **params), axis=-1)

    def jax_llf(p):
        params = dict(zip(pnames, p))
        try:
            return -jaxdist.logpdf(x, **params).sum()
        except AttributeError:
            return -jaxdist.logpmf(x, **params).sum()

    res = differentiate.hessian(llf, p, tolerances={'atol': 1e-10},
                                initial_step=0.1)

    H = jax.hessian(jax_llf)
    ref = np.asarray(H(jax.numpy.asarray(p)))
    np.testing.assert_allclose(res.ddf, ref, rtol=1e-6, atol=1e-6)
```

The tricky (mostly just annoying, hopefully) part is restricting function evaluations to the domain of parameters and values of the parameters that are consistent with the data (i.e. PDF/PMF of data is not 0).

I have a PR (gh-21638) that would allow `jacobian` to control which of the one- or two-sided finite finite difference approximations is used for each variable. This would need to be extended to `hessian` (easy, I think) and then the logic for setting this (and maybe the `initial_step`) based on the particulars of the distribution would need to be added (annoying). An intermediate step might be to wrap `jacobian` to accept `bounds` for each variable; which might make the logic easier to follow overall.

A different approach would be to add logic in `differentiate` (soon to be `derivative`) that automatically chooses the step direction based on whether it encounters NaNs in its initial function evaluations. This would probably be much easier, at the expense of a few wasted function evaluations. (Fine, IMO. Especially after not having this functionality at all many years after the need was first identified.)

Yet another approach would be to add an autodiff simple, private autodiff array wrapper. I understand why we have not wanted to get into the public autodiff business, but it makes it challenging sometimes when we need derivatives internally.

The other question is what to do about parameters that are inherently discrete. Above, `nbinom` is extended to all positive `n`, but that's not the case for all discrete distributions (e.g. `binom`).
